### ISSUE TITLE
Field and Constructor instances in Data.Generic.Labels

### ISSUE THREAD

**Issue by arybczak** [kcsongor/generic-lens#81]
Can you explain why these are incoherent?

I think if it works as expected, it does only by accident. If any of the `s ~ t` or `a ~ b` doesn't hold, then `Field name s t a b` is picked, but otherwise arbitrary instance from `Field name s s a a` and `Field name s t a b` will be picked. Having

```haskell
instance {-# OVERLAPPABLE #-} HasField name s t a b => Field name s t a b where
  fieldLens = field @name

instance HasField' name s a => Field name s s a a where
  fieldLens = field' @name
```
or potentially
```haskell
instance HasField name s t a b => Field name s t a b where
  fieldLens = field @name

instance {-# OVERLAPPING #-} HasField' name s a => Field name s s a a where
  fieldLens = field' @name
```
would make instance choice deterministic.

Can you also please explain how these instances differ? Why can't just the one for more general case be used? Unfortunately I can't seem to be able to decipher this from their definitions.

**Comment by kcsongor** [comment]
The `HasField'` class in its implementation is identical to the `HasField` one, but because the types are known not to change, it does less computation, and thus compiles faster. There are also cases where not changing types is more robust (when the parameter appears at a nominal role), but detecting these cases is not possible due to the rather second class role system of GHC, as far as I'm aware.
Long story short, there's no operational difference in these functions, and it's not the end of the world if the more general one is selected. As such, this 'by accident' behaviour is fine. Having the `OVERLAPPING` pragma would mean that in cases where `t` and `b` are not known, we would get a type error, whereas with incoherent instances, it just defaults to the general version.

If you have a better suggestion, I would be happy to hear it though!

**Comment by arybczak** [comment]
Thanks for the explanation!

What about marking only the more general instance as INCOHERENT?

From https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html#extension-IncoherentInstances:
> - If exactly one non-incoherent candidate remains, select it. If all remaining candidates are incoherent, select an arbitrary one. Otherwise the search fails (i.e. when more than one surviving candidate is not incoherent).
> - If the selected candidate (from the previous step) is incoherent, the search succeeds, returning that candidate.

I think that marking only the more general instance as INCOHERENT would work and make the choice deterministic.

**Comment by kcsongor** [comment]
Yes, indeed that sounds like a good idea - thanks for pointing it out!
